File: java/data_structures/priority_queue/builtin/PriorityQueueBuiltin.java
package priority_queue.builtin;

import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Priority Queue implemented with Java's built-in PriorityQueue.
 *
 * Features:
 * - Min or Max behavior (configurable via constructor).
 * - Stable among equal priorities (FIFO) using a monotonic counter.
 *
 * API:
 *  - push(value, priority) : O(log n)
 *  - pop()                 : O(log n)
 *  - peek()                : O(1)
 *  - size(), isEmpty()     : O(1)
 *  - clear()               : O(n)
 */
public class PriorityQueueBuiltin<T> {

    private static final class Node<T> {
        final int priority;
        final long order; 
        final T value;

        Node(int priority, long order, T value) {
            this.priority = priority;
            this.order = order;
            this.value = value;
        }
    }

    private final PriorityQueue<Node<T>> pq;
    private final AtomicLong counter = new AtomicLong(0);


    public PriorityQueueBuiltin() {
        this(true);
    }


    public PriorityQueueBuiltin(boolean isMin) {
        Comparator<Node<T>> cmp = (a, b) -> {
            if (a.priority != b.priority) {
                return isMin
                        ? Integer.compare(a.priority, b.priority)      // min: lower first
                        : Integer.compare(b.priority, a.priority);     // max: higher first
            }
            return Long.compare(a.order, b.order);
        };
        this.pq = new PriorityQueue<>(cmp);
    }

    public void push(T value, int priority) {
        long order = counter.getAndIncrement(); // tie-breaker for stability
        pq.add(new Node<>(priority, order, value));
    }

    public T pop() {
        Node<T> n = pq.poll();
        if (n == null) throw new IllegalStateException("pop from an empty priority queue");
        return n.value;
    }

    public T peek() {
        Node<T> n = pq.peek();
        if (n == null) throw new IllegalStateException("peek from an empty priority queue");
        return n.value;
    }

    /* Current number of elements. */
    public int size() {
        return pq.size();
    }

    /* True if the queue has no elements. */
    public boolean isEmpty() {
        return pq.isEmpty();
    }

    /* Remove all elements. */
    public void clear() {
        pq.clear();
    }
}

File: csharp/data_structure/priority_queue/builtin/PriorityQueueBuiltin.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace DataStructure.PriorityQueue.Builtin;

/// <summary>
/// Priority Queue implemented using .NET's built-in PriorityQueue&lt;TElement, TPriority&gt;.
///
/// Features:
/// - Supports Min or Max behavior (configurable via constructor)
/// - Stable among equal priorities (FIFO) using a monotonic counter
///
/// Time Complexity:
/// - Push(): O(log n)
/// - Pop():  O(log n)
/// - Peek(): O(1)
/// - Size(), IsEmpty(): O(1)
/// </summary>
public class PriorityQueueBuiltin<T>
{
    private readonly PriorityQueue<QueueItem<T>, (int priority, long order)> _pq;
    private readonly bool _isMin;
    private long _counter;

    public PriorityQueueBuiltin(bool isMin = true)
    {
        _isMin = isMin;
        _pq = new PriorityQueue<QueueItem<T>, (int priority, long order)>();
        _counter = 0;
    }

    public void Push(T value, int priority)
    {
        long order = Interlocked.Increment(ref _counter);

        // For stability and direction:
        // - Min-queue: smaller priority first, then smaller order first
        // - Max-queue: invert priority to negative so that "larger" comes first
        var effectivePriority = _isMin ? priority : -priority;
        var item = new QueueItem<T>(value, priority, order);
        _pq.Enqueue(item, (effectivePriority, order));
    }

    public T Pop()
    {
        if (_pq.Count == 0)
            throw new InvalidOperationException("Pop from an empty priority queue");

        return _pq.Dequeue().Value;
    }

    public T Peek()
    {
        if (_pq.Count == 0)
            throw new InvalidOperationException("Peek from an empty priority queue");

        return _pq.Peek().Value;
    }

    public int Size() => _pq.Count;
    public bool IsEmpty() => _pq.Count == 0;
    public void Clear() => _pq.Clear();

    // --- Internal node class ---
    private sealed class QueueItem<TValue>
    {
        public TValue Value { get; }
        public int OriginalPriority { get; }
        public long Order { get; }

        public QueueItem(TValue value, int priority, long order)
        {
            Value = value;
            OriginalPriority = priority;
            Order = order;
        }

        public override string ToString() => $"({OriginalPriority}, {Order}, {Value})";
    }
}
